src/engine/types/actions.rs

```diff
@@ -575,38 +575,44 @@ impl Display for DeleteAction<Key> {
 }
 
 /// An action which may delete multiple things.
+/// This action may also cause other values to require updating.
 #[derive(Debug, PartialEq, Eq)]
-pub struct SetDeleteAction<T> {
+pub struct SetDeleteAction<T, U> {
     changed: Vec<T>,
+    updated: Vec<U>,
 }
 
-impl<T> SetDeleteAction<T> {
-    pub fn new(changed: Vec<T>) -> Self {
-        SetDeleteAction { changed }
+impl<T, U> SetDeleteAction<T, U> {
+    pub fn new(changed: Vec<T>, updated: Vec<U>) -> Self {
+        assert!(!changed.is_empty() || updated.is_empty());
+        SetDeleteAction { changed, updated }
     }
 
     pub fn empty() -> Self {
-        SetDeleteAction { changed: vec![] }
+        SetDeleteAction {
+            changed: vec![],
+            updated: vec![],
+        }
     }
 }
 
-impl<T> EngineAction for SetDeleteAction<T> {
-    type Return = Vec<T>;
+impl<T, U> EngineAction for SetDeleteAction<T, U> {
+    type Return = (Vec<T>, Vec<U>);
 
     fn is_changed(&self) -> bool {
         !self.changed.is_empty()
     }
 
-    fn changed(self) -> Option<Vec<T>> {
+    fn changed(self) -> Option<(Vec<T>, Vec<U>)> {
         if self.changed.is_empty() {
             None
         } else {
-            Some(self.changed)
+            Some((self.changed, self.updated))
         }
     }
 }
 
-impl Display for SetDeleteAction<FilesystemUuid> {
+impl<U> Display for SetDeleteAction<FilesystemUuid, U> {
```

Comment:
I think erring on the side of specificity is good for these `impl`s so using `FilesystemUuid` might be better. Otherwise, we have a frequent problem of overlapping impls. That would also allow us to add to the log message indicating what UUIDs had their origin updated.